### PR TITLE
TCP G: Add support for fixed-width types: times, datetimes, IPs, UUID, BFloat16

### PR DESCRIPTION
**Pull request by alex-clickhouse**
# TCP Epic G — Fixed-width types (read + write)

Completes the fixed-width type family for the native/TCP client (`ClickHouse.Driver.Tcp`),
so the columnar read/write path now covers the everyday scalar types a real workload uses.
Stacked on `tcp/epic-f-basic-inserts`.

Every type has both a read (block → column) and a write (column → block) path, is registered
in `ColumnCodecRegistry`, and is exercised by a live INSERT→SELECT round-trip against a real
server in addition to unit tests.

## Types added

| ClickHouse | CLR (`Values`) | Notes |
|---|---|---|
| `Float32` / `Float64` | `float` / `double` | direct blittable reinterpret |
| `BFloat16` | `float` | widened on read (top 16 bits), narrowed on write |
| `Bool` | `bool` | 1 byte, canonical 0/1 |
| `Date` / `Date32` | `DateOnly` | UInt16 / Int32 days since epoch, range-checked |
| `DateTime` / `DateTime('tz')` | `uint` | raw epoch seconds; `GetDateTimeOffset`/`ToDateTimeOffsets` projection (per-instant offset from the type's tz or the session tz) |
| `DateTime64(scale[,'tz'])` | `long` | raw counts, exact at every scale (incl. sub-100 ns 8/9); lossy `DateTimeOffset` projection; scale/tz held once on the column |
| `Time` / `Time64(scale)` | `int` / `long` | raw seconds / scaled counts; lossy `TimeSpan` projection (`GetTimeSpan`/`ToTimeSpans`) |
| `Interval*` | `long` | underlying Int64 count; unit kept in the type name |
| `UUID` | `Guid` | ClickHouse's byte-swapped halves — see the SIMD note below |
| `IPv4` / `IPv6` | `IPAddress` | v4 byte-reversed, v6 network order |
| `Enum8` / `Enum16` | `sbyte` / `short` | raw ordinal surfaced; `'label'=ordinal` map retained as metadata |
| `Decimal32/64` | `decimal` | |
| `Decimal(P,S)` / `Decimal128/256` | `ClickHouseDecimal` | Int256-backed, for values beyond `decimal`'s range |
| `Nothing` | `null` placeholder | base codec; `Nullable(Nothing)` deferred to the composite epic |
| `String` | `string` / raw `byte[]` | reworked — see below |

## Performance-relevant design choices

The theme throughout: **read into pooled storage, reinterpret zero-copy, and convert on demand** —
never eagerly materialize a converted array we might not fully consume.

- **Pooled, borrowed column buffers.** Column backing arrays are rented from `ArrayPool` and
  returned on `Dispose`; a column (and its `Values`/spans) is valid only for its block's iteration.
  This removed the per-block conversion-array allocation (previously a fresh `T[]` per column, per block).

- **Zero-copy reinterpret, raw wire value as the default surface.** The temporal columns
  (`DateTimeColumn`/`DateTime64Column`/`TimeColumn`/`Time64Column`) keep the raw wire bytes as their
  storage and expose them directly as `Values` via `MemoryMarshal.Cast` — `uint` epoch seconds, `int`
  seconds, or `long` counts. The raw count is the exact wire value at every scale (including the sub-100 ns
  `DateTime64`/`Time64` scales 8/9 that no `DateTimeOffset`/`TimeSpan` can hold). A caller that wants a
  calendar/duration value asks for it: `GetDateTimeOffset`/`ToDateTimeOffsets` and
  `GetTimeSpan`/`ToTimeSpans` project on demand (lossy for scales > 7, documented). The scale and
  timezone — identical for every row — live **once on the column** (`Scale`/`TimeZone`), not per value,
  so nothing is paid per stored row. This replaced the earlier project-local `ClickHouseDateTime64`
  struct, which is removed.

- **Per-instant offset resolution (no fixed-offset fast path).** `DateTimeOffset` projection resolves
  the offset from the instant via `TimeZoneInfo.ConvertTime`, so both DST transitions **and** historical
  base-offset changes are honored — `SupportsDaylightSavingTime == false` does not imply a single offset
  for all time.

- **SIMD UUID shuffle.** ClickHouse stores a UUID as two little-endian 64-bit halves — a fixed 16-byte
  permutation from a `Guid`'s layout. That shuffle is now one `Vector128.Shuffle` (`PSHUFB`/`TBL`) per
  value, applied straight between the wire blob and the `Guid[]` memory (no `new Guid(...)` parse, no
  scratch), with a scalar fallback. Benchmarked ~9–10× the scalar byte-by-byte shuffle, zero-alloc; the
  read path realizes it fully. Relies on the same little-endian assumption the other reinterpret paths make.

- **`StringColumn` keeps raw bytes, decodes on demand.** One pooled blob + per-row offsets; a caller
  picks `GetBytes(row)` (zero-copy), `GetString(row, encoding)`, or the default UTF-8 `Values` view.
  Byte-oriented / non-UTF-8 data is fully accessible without forcing an eager `string[]` on read.

- **`ResolveContext` threaded as a value, not connection state.** The timezone context for a read is
  computed once per operation (from the query's `session_timezone` or the handshake tz) and passed down
  the read path as a parameter — no mutable field to leak one query's timezone into the next, and the
  shared codec registry stays free of any per-connection cache.

- **Removed per-codec little-endian guards.** The reinterpret paths assume little-endian; the redundant
  per-call checks were dropped in favor of a single assertion at connection open (tracked follow-up),
  keeping the hot read/write paths branch-free.

## Write paths

Each temporal codec writes the raw value verbatim (`uint`/`int`/`long`, assumed already at the column's
scale) and also accepts the convenience CLR types for ergonomics: `DateTime`/`DateTimeOffset` for
`DateTime`/`DateTime64`, and `TimeSpan` for `Time`/`Time64` (range-checked, truncated to scale).

## Timezones

`DateTime`/`DateTime64` project to `DateTimeOffset` with the offset resolved **per instant** (so DST
transitions are honored), from the type string's tz argument, else the session timezone, else UTC.
Timezone resolution also handles ClickHouse's synthetic `Fixed/UTC±HH:MM:SS` offset names (parsed into a
custom `TimeZoneInfo`), and a query's `session_timezone` setting is applied as the presentation timezone
for timezone-less columns.

## Wide numerics

`ClickHouseDecimal` — an Int256 mantissa + scale — carries `Decimal128`/`Decimal256` values beyond
`System.Decimal`'s range, mirroring how wide integers are handled. (`DateTime64` no longer needs a
project-local carrier: the raw `long` count is the lossless surface.)

## Testing

Live INSERT→SELECT round-trips (`InsertRoundTripCase`) cover every type end-to-end against a real
server — including a `DateTime64(9)` nanosecond round-trip and `Fixed/UTC` / `session_timezone` cases —
backed by per-codec unit tests for edge cases (negatives, boundaries, byte-order vectors, precision/scale
validation, non-UTF-8 strings, a UUID fuzz round-trip). ~85%+ line coverage on the new code.

### CODE REVIEW

File: ClickHouse.Driver.Tcp.Tests/Types/DateTime64ColumnCodecTests.cs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+[TestFixture]
+public class DateTime64ColumnCodecTests
+{
+    private static DateTime64ColumnCodec Codec(string type, string tz = null) => DateTime64ColumnCodec.Create(TypeParser.Parse(type), tz);
+
+    [TestCase("DateTime64(0)")]
+    [TestCase("DateTime64(3)")]
+    [TestCase("DateTime64(6)")]
+    [TestCase("DateTime64(7)")]
```
Comment (kavirajk) [inline ClickHouse.Driver.Tcp.Tests/Types/DateTime64ColumnCodecTests.cs:19]: Can you add Case 8 and 9 here? 

I know we truncate because .NET doesn't support nano second prescition. 

Just to avoid any future regression

File: ClickHouse.Driver.Tcp/Types/Codecs/DateTimeZones.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    /// <param name="serverTimezone">The session's timezone, or null/empty when unknown (e.g. the write path).</param>
+    /// <returns>The resolved timezone info.</returns>
+    /// <exception cref="FormatException">The named timezone is not known to the platform.</exception>
+    public static TimeZoneInfo Resolve(string explicitTimezone, string serverTimezone)
+    {
+        string id = !string.IsNullOrEmpty(explicitTimezone) ? explicitTimezone
+            : !string.IsNullOrEmpty(serverTimezone) ? serverTimezone
+            : null;
+
+        if (id is null)
```
Comment (kavirajk) [inline ClickHouse.Driver.Tcp/Types/Codecs/DateTimeZones.cs:37]: can we make sure we have tests for this case of UTC if no server timezone.

File: ClickHouse.Driver.Tcp/Types/Codecs/DateColumnCodec.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+using System.Threading.Tasks;
+using ClickHouse.Driver.Tcp.Protocol;
+
+namespace ClickHouse.Driver.Tcp.Types.Codecs;
+
+/// <summary>
+/// A codec for the ClickHouse <c>Date</c> column: a little-endian <c>UInt16</c> day count since the Unix epoch
+/// (1970-01-01), surfaced as a <see cref="DateOnly"/>. The representable range is 1970-01-01 to 2149-06-06.
+/// </summary>
+internal sealed class DateColumnCodec : IColumnCodec
```
Comment (kavirajk) [inline ClickHouse.Driver.Tcp/Types/Codecs/DateColumnCodec.cs:13]: Can we have some tests to remove the "time" portion of "DateTime" to avoid interrupting it in-correctly. When user only needs "Date" not the time.

NOTE: Also needs clearing timezone.

I had some real bug in go before

File: ClickHouse.Driver.Tcp/Types/Codecs/BFloat16ColumnCodec.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace ClickHouse.Driver.Tcp.Types.Codecs;
+
+/// <summary>
+/// A codec for the ClickHouse <c>BFloat16</c> column: a 16-bit "brain float" (the top 16 bits of an IEEE-754
+/// <c>Float32</c> — same 8-bit exponent, 7-bit mantissa) surfaced as a widened <see cref="float"/>. There is no
+/// BCL <c>BFloat16</c> type, so a value is widened on read (shift the 16 bits into the high half of a 32-bit
+/// float) and narrowed on write (drop the low 16 bits, losing mantissa precision).
+/// </summary>
+internal sealed class BFloat16ColumnCodec : IColumnCodec
```
Comment (alex-clickhouse) [inline ClickHouse.Driver.Tcp/Types/Codecs/BFloat16ColumnCodec.cs:15]: Double check transpose wire format

**Review by kavirajk — COMMENTED** [review]
LGTM. Just few clarification and tests coverage for some cases

File: ClickHouse.Driver.Tcp/Types/Codecs/DateTimeZones.cs (hunk 2 above)
Comment (alex-clickhouse) [reply ClickHouse.Driver.Tcp/Types/Codecs/DateTimeZones.cs:37]: done

File: ClickHouse.Driver.Tcp.Tests/Types/DateTime64ColumnCodecTests.cs (hunk 1 above)
Comment (alex-clickhouse) [reply ClickHouse.Driver.Tcp.Tests/Types/DateTime64ColumnCodecTests.cs:19]: done

File: ClickHouse.Driver.Tcp/Types/Codecs/DateColumnCodec.cs (hunk 3 above)
Comment (alex-clickhouse) [reply ClickHouse.Driver.Tcp/Types/Codecs/DateColumnCodec.cs:13]: We use DateOnly here, so there's no time portion.

**Review by kavirajk — APPROVED** [review]
LGTM
